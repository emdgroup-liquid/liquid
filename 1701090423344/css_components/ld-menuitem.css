:host{display:block}.ld-menuitem .ld-menuitem__button{display:grid;width:100%}