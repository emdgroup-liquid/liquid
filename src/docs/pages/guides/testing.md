---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

