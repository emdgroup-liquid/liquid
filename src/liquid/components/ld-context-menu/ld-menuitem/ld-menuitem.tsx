import {
  Component,
  Element,
  Host,
  h,
  Prop,
  State,
  Method,
  Event,
  EventEmitter,
} from "@stencil/core";
import { cloneAttributes } from "../../../utils/cloneAttributes";

type Mode = "highlight" | "danger" | "neutral";

const modeMap = new Map<Mode, string>([
  ["danger", "danger-ghost"],
  ["highlight", "ghost"],
  ["neutral", "neutral-ghost"],
]);

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part listitem - `li` element wrapping the `ld-button` element
 * @part button - `ld-button` element wrapping the default slot
 */
@Component({
  tag: "ld-menuitem",
  styleUrl: "ld-menuitem.css",
  shadow: true,
})
export class LdMenuitem implements InnerFocusable {
  @Element() el: HTMLLdMenuitemElement;
  private attributesObserver: MutationObserver;
  private buttonRef?: HTMLLdButtonElement;

  /** Disabled state of the menu item. */
  @Prop() disabled?: boolean;

  /**
   * Transforms the menu item to an anchor element.
   * See [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-href)
   * for more information on the `href` attribute.
   */
  @Prop() href?: HTMLLdButtonElement["href"];

  /** Tab index of the menu item. */
  @Prop() ldTabindex?: number;

  /** Prevent closing of the context menu on click. */
  @Prop() preventClose?: boolean;

  /** Display mode. */
  @Prop() mode?: Mode = "neutral";

  /**
   * Size of the menu item.
   * @internal
   */
  @Prop() size?: "sm" | "lg";

  /**
   * The `target` attributed can be used in conjunction with the `href` attribute.
   * See [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target)
   * for more information on the `target` attribute.
   */
  @Prop() target?: HTMLLdButtonElement["target"];

  @State() clonedAttributes: Record<string, string>;

  /** Sets focus on the anchor or button */
  @Method()
  async focusInner() {
    this.buttonRef?.focusInner();
  }

  /**
   * @internal
   * Emitted on menu item click if preventClose prop is not truethy.
   */
  @Event() ldclosetooltip: EventEmitter;

  private handleClick = (ev: MouseEvent) => {
    if (this.preventClose) return;
    this.ldclosetooltip.emit(ev);
  };

  componentWillLoad() {
    this.attributesObserver = cloneAttributes.call(this, [
      "ld-tabindex",
      "mode",
      "size",
    ]);
  }

  disconnectedCallback() {
    /* istanbul ignore if */
    if (this.attributesObserver) this.attributesObserver.disconnect();
  }

  render() {
    return (
      <Host>
        <li class="ld-menuitem" part="listitem" role="menuitem">
          <ld-button
            {...this.clonedAttributes}
            class="ld-menuitem__button"
            disabled={this.disabled}
            href={this.href}
            iconOnly={false}
            justifyContent="start"
            ldTabindex={this.ldTabindex}
            mode={modeMap.get(this.mode)}
            onClick={this.handleClick}
            part="focusable button"
            ref={(element) => (this.buttonRef = element)}
            size={this.size}
            target={this.target}
            type="button"
          >
            <slot></slot>
          </ld-button>
        </li>
      </Host>
    );
  }
}
