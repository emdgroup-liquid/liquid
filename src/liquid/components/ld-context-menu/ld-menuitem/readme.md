---
eleventyNavigation:
  key: Menuitem
  parent: Context Menu
layout: layout.njk
title: Menuitem
permalink: components/ld-context-menu/ld-menuitem/
---

<link rel="stylesheet" href="css_components/ld-menu.css">
<link rel="stylesheet" href="css_components/ld-menuitem.css">
<link rel="stylesheet" href="css_components/ld-icon.css">
<link rel="stylesheet" href="css_components/ld-button.css">

# ld-menuitem

The `ld-menuitem` component is a subcomponent for `ld-context-menu` / `ld-menu` and is meant to be used in the slot of either the [`ld-context-menu`](..), the [`ld-menu`](../ld-menu) or the [`ld-menuitem-group`](../ld-menuitem-group) component.

---

## Examples

### Default

{% example '{ "opened": true }' %}
<ld-menu>
  <ld-menuitem>Menu item</ld-menuitem>
</ld-menu>

<!-- React component -->

<LdMenu>
  <LdMenuitem>Menu item</LdMenuitem>
</LdMenu>

<!-- CSS component -->

<ul class="ld-menu" role="menu">
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
      Menu item
    </button>
  </li>
</ul>
{% endexample %}

### Highlight

{% example '{ "opened": true }' %}
<ld-menu>
  <ld-menuitem mode="highlight">Menu item</ld-menuitem>
</ld-menu>

<!-- React component -->

<LdMenu>
  <LdMenuitem mode="highlight">Menu item</LdMenuitem>
</LdMenu>

<!-- CSS component -->

<ul class="ld-menu" role="menu">
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost">
      Menu item
    </button>
  </li>
</ul>
{% endexample %}

### Danger

{% example '{ "opened": true }' %}
<ld-menu>
  <ld-menuitem mode="danger">Menu item</ld-menuitem>
</ld-menu>

<!-- React component -->

<LdMenu>
  <LdMenuitem mode="danger">Menu item</LdMenuitem>
</LdMenu>

<!-- CSS component -->

<ul class="ld-menu" role="menu">
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--danger-ghost">
      Menu item
    </button>
  </li>
</ul>
{% endexample %}

### As link

{% example '{ "opened": true }' %}
<ld-menu>
  <ld-menuitem href="#" target="_blank">Menu item</ld-menuitem>
</ld-menu>

<!-- React component -->

<LdMenu>
  <LdMenuitem href="#" target="_blank">Menu item</LdMenuitem>
</LdMenu>

<!-- CSS component -->

<ul class="ld-menu" role="menu">
  <li class="ld-menuitem"  role="menuitemradio">
    <a class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost" href="#" target="_blank" rel="noreferrer noopener">
      Menu item
    </a>
  </li>
</ul>
{% endexample %}

<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description                                                                                                                                                                                                           | Type                                         | Default     |
| ------------ | ------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------- | ----------- |
| `href`       | `href`        | Transforms the menu item to an anchor element. See [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-href) for more information on the `href` attribute.                                    | `string`                                     | `undefined` |
| `ldTabindex` | `ld-tabindex` | Tab index of the menu item.                                                                                                                                                                                           | `number`                                     | `undefined` |
| `mode`       | `mode`        | Display mode.                                                                                                                                                                                                         | `"danger" \| "highlight" \| "neutral"`       | `'neutral'` |
| `target`     | `target`      | The `target` attributed can be used in conjunction with the `href` attribute. See [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target) for more information on the `target` attribute. | `"_blank" \| "_parent" \| "_self" \| "_top"` | `undefined` |


## Methods

### `focusInner() => Promise<void>`

Sets focus on the anchor or button

#### Returns

Type: `Promise<void>`




## Dependencies

### Depends on

- [ld-button](../../ld-button)

### Graph
```mermaid
graph TD;
  ld-menuitem --> ld-button
  style ld-menuitem fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
