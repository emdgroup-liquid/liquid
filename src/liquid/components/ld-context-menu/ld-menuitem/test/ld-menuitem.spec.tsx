import { newSpecPage } from "@stencil/core/testing";
import { LdMenuitem } from "../ld-menuitem";
import {
  clearTriggerableMutationObservers,
  getTriggerableMutationObservers,
} from "../../../../utils/mutationObserver";

describe("ld-menuitem", () => {
  afterEach(clearTriggerableMutationObservers);

  it("renders default", async () => {
    const page = await newSpecPage({
      components: [LdMenuitem],
      html: `<ld-menuitem>Menu item</ld-menuitem>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("maps mode prop", async () => {
    const page = await newSpecPage({
      components: [LdMenuitem],
      html: `<ld-menuitem mode="neutral">Menu item 1</ld-menuitem>
      <ld-menuitem mode="highlight">Menu item 2</ld-menuitem>
      <ld-menuitem mode="danger">Menu item 3</ld-menuitem>`,
    });
    expect(page.body).toMatchSnapshot();
  });

  it("renders other known props", async () => {
    const page = await newSpecPage({
      components: [LdMenuitem],
      html: `<ld-menuitem
        disabled
        href="#"
        ld-tabindex="1"
        size="sm"
        target="_blank"
      >Menu item</ld-menuitem>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("clones attributes", async () => {
    const page = await newSpecPage({
      components: [LdMenuitem],
      html: `<ld-menuitem data-attribute="example">Menu item</ld-menuitem>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("updates cloned attributes", async () => {
    const page = await newSpecPage({
      components: [LdMenuitem],
      html: `<ld-menuitem data-attribute="example">Menu item</ld-menuitem>`,
    });

    page.root.setAttribute("data-attribute", "test");
    getTriggerableMutationObservers()[0].trigger([
      { attributeName: "data-attribute" },
    ]);
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it("allows to set inner focus", async () => {
    const { root } = await newSpecPage({
      components: [LdMenuitem],
      html: `<ld-menuitem data-attribute="example">Menu item</ld-menuitem>`,
    });
    const button = root.shadowRoot.querySelector("ld-button");

    button.focusInner = jest.fn();
    await root.focusInner();

    expect(button.focusInner).toHaveBeenCalled();
  });
});
