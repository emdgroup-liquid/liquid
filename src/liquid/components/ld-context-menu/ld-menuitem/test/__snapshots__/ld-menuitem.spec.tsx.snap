// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-menuitem clones attributes 1`] = `
<ld-menuitem data-attribute="example">
  <mock:shadow-root>
    <li class="ld-menuitem" role="menuitem">
      <ld-button class="ld-menuitem__button" data-attribute="example" justifycontent="start" mode="neutral-ghost" type="button">
        <slot></slot>
      </ld-button>
    </li>
  </mock:shadow-root>
  Menu item
</ld-menuitem>
`;

exports[`ld-menuitem maps mode prop 1`] = `
<body>
  <ld-menuitem mode="neutral">
    <mock:shadow-root>
      <li class="ld-menuitem" role="menuitem">
        <ld-button class="ld-menuitem__button" justifycontent="start" mode="neutral-ghost" type="button">
          <slot></slot>
        </ld-button>
      </li>
    </mock:shadow-root>
    Menu item 1
  </ld-menuitem>
  <ld-menuitem mode="highlight">
    <mock:shadow-root>
      <li class="ld-menuitem" role="menuitem">
        <ld-button class="ld-menuitem__button" justifycontent="start" mode="ghost" type="button">
          <slot></slot>
        </ld-button>
      </li>
    </mock:shadow-root>
    Menu item 2
  </ld-menuitem>
  <ld-menuitem mode="danger">
    <mock:shadow-root>
      <li class="ld-menuitem" role="menuitem">
        <ld-button class="ld-menuitem__button" justifycontent="start" mode="danger-ghost" type="button">
          <slot></slot>
        </ld-button>
      </li>
    </mock:shadow-root>
    Menu item 3
  </ld-menuitem>
</body>
`;

exports[`ld-menuitem renders default 1`] = `
<ld-menuitem>
  <mock:shadow-root>
    <li class="ld-menuitem" role="menuitem">
      <ld-button class="ld-menuitem__button" justifycontent="start" mode="neutral-ghost" type="button">
        <slot></slot>
      </ld-button>
    </li>
  </mock:shadow-root>
  Menu item
</ld-menuitem>
`;

exports[`ld-menuitem renders other known props 1`] = `
<ld-menuitem disabled="" href="#" ld-tabindex="1" size="sm" target="_blank">
  <mock:shadow-root>
    <li class="ld-menuitem" role="menuitem">
      <ld-button class="ld-menuitem__button" disabled="" href="#" justifycontent="start" ldtabindex="1" mode="neutral-ghost" size="sm" target="_blank" type="button">
        <slot></slot>
      </ld-button>
    </li>
  </mock:shadow-root>
  Menu item
</ld-menuitem>
`;

exports[`ld-menuitem updates cloned attributes 1`] = `
<ld-menuitem data-attribute="test">
  <mock:shadow-root>
    <li class="ld-menuitem" role="menuitem">
      <ld-button class="ld-menuitem__button" data-attribute="test" justifycontent="start" mode="neutral-ghost" type="button">
        <slot></slot>
      </ld-button>
    </li>
  </mock:shadow-root>
  Menu item
</ld-menuitem>
`;
